### PR TITLE
Prepare group menu for a11y

### LINKED ISSUE #12089: Use buttons instead of links in ConfiguratorGroupMenuComponent
Correlate to #10525

We're should not use anchor elements for things that are sematically buttons (they don't have `href` attribute, but invoke JS logic). But it's the case for `ConfiguratorGroupMenuComponent`. For example:
```typescript
        <a
          class="cx-menu-back-link"
          role="link"
          tabindex="0"
          (click)="navigateUp()"
          (keyup)="navigateUpOnEnter($event)"
        >
```
That forces us to implement custom event handler (keyup), which is a redundant logic.
Or to implement redundant custom event handler to avoid scrolling down for spacebar `(keydown)="preventScrollingOnSpace($event)"`

Instead we should use a `<button>` which is semantically correct. Btw. it also supports space bar natively. So we can also get rid off all unnecessary custom event handlers.

Changing DOM elements from `<a>` to `<button>` is a breaking change in the DOM. So should be done in a major release.

Make possible to navigation via Arrow keys within the group menu: [w3.org](https://www.w3.org/TR/2019/NOTE-wai-aria-practices-1.1-20190814/examples/tabs/tabs-2/tabs.html)

### PR DESCRIPTION
**Pull request by Larisa-Staroverova**

### CODE REVIEW

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts [hunk 1]
```diff
@@ -16,6 +16,7 @@ export class ConfiguratorStorefrontUtilsService {
   constructor(
     protected configuratorGroupsService: ConfiguratorGroupsService,
     @Inject(PLATFORM_ID) protected platformId: any,
+    @Inject(DOCUMENT) protected document,
```
Comment (Platonn) [inline feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts]: nitpick: is this change needed?

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+import { DOCUMENT, isPlatformBrowser } from '@angular/common';
+import { BREAKPOINT, BreakpointService } from '@spartacus/storefront';
+import { take } from 'rxjs/operators';
+
+@Injectable({ providedIn: 'root' })
+export class ConfiguratorGroupMenuService {
+  constructor(
+    @Inject(PLATFORM_ID) protected platformId: any,
+    @Inject(DOCUMENT) protected document,
+    protected breakpointService?: BreakpointService
```
Comment (Platonn) [inline feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts]: Why this dependency is optional? Typo?

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.ts [hunk 3]
```diff
@@ -290,4 +285,56 @@ export class ConfiguratorGroupMenuComponent {
...
+   * @param {KeyboardEvent} event - Keyboard event
+   */
+  switchGroupOnArrowPress(
+    event: KeyboardEvent,
+    groupIndex: number,
+    group?: Configurator.Group
+  ): void {
+    if (event.code === 'ArrowUp' || event.code === 'ArrowDown') {
+      this.configGroupMenuService.switchGroupOnArrowPress(event, groupIndex);
+    } else if (event.code === 'ArrowLeft' || event.code === 'ArrowRight') {
```
Comment (Platonn) [inline feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.ts]: Why both on arrow left an right we perform going deeper in the tree or going up the tree? 
[opinion] I'd expect arrow left to always let me navigate out (up) in the tree. But arrow right to navigate deeper (if possible). But it's just a personal opinion. Please verify standards with @developpeurweb. 

If I'm right above, then I perhaps we should swap the meaning of arrow left and right for RTL storefronts. Please verify with @developpeurweb too.

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.html [hunk 4]
```diff
@@ -1,66 +1,60 @@
...
     <ng-container *ngIf="displayedGroups$ | async as groups">
       <ng-container *ngIf="currentGroup$ | async as currentGroup">
-        <ng-container *ngFor="let group of groups">
-          <li
+        <ng-container *ngFor="let group of groups; let groupIndex = index">
+          <ng-container *ngIf="displayedParentGroup$ | async as parentGroup">
+            <button
+              *ngIf="parentGroup !== null && groupIndex === 0"
+              class="cx-menu-back"
+              id="back-button"
```
Comment (Platonn) [inline feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.html]: It would be good to avoid arbitrary global ID. Can we use other special marker for the back button? Class name?

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
+import { Inject, Injectable, PLATFORM_ID } from '@angular/core';
+import { DOCUMENT, isPlatformBrowser } from '@angular/common';
+import { BREAKPOINT, BreakpointService } from '@spartacus/storefront';
+import { take } from 'rxjs/operators';
+
+@Injectable({ providedIn: 'root' })
+export class ConfiguratorGroupMenuService {
+  constructor(
+    @Inject(PLATFORM_ID) protected platformId: any,
```
Comment (Platonn) [inline feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts]: Is this check _really_ needed in every method? `isPlatformBrowser(this.platformId)`? I hope not. That would simplify a bit the methods.

I hope we're safe in SSR, when you use `this.document` (Angular Injectable `DOCUMENT` wrapper) instead of the global `document` object.

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+  }
+
+  protected updateCurrentGroupIndex(
+    currentGroupIndex: number,
+    focusedGroupIndex: number
+  ): number {
+    return focusedGroupIndex !== currentGroupIndex
+      ? focusedGroupIndex
+      : currentGroupIndex;
+  }
```
Comment (Platonn) [inline feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts]: This method doesn't update the current group index. But that's what I understand on the first sight from the method name.

Could you please change the name to be more descriptive? You might also consider to add a JSDoc comment for the method.

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.ts [hunk 7]
```diff
@@ -290,4 +285,56 @@ export class ConfiguratorGroupMenuComponent {
...
+      group.subGroups.forEach((subGroup) => {
+        if (this.isGroupSelected(subGroup.id, currentGroupId)) {
+          isCurrentGroupFound = true;
+        }
+      });
+    }
+    return isCurrentGroupFound;
+  }
+
+  setTabIndex(currentGroupId: string, group: Configurator.Group): number {
```
Comment (Platonn) [inline feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.ts]: nitpick: It doesn't set any value, but returns a value. Let's rename it to getTabIndex

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+            selector = 'main' + selector;
+          } else {
+            selector = '.navigation' + selector;
+          }
+        });
+      return document.querySelectorAll(selector);
+    }
+  }
+
+  protected getFocusedGroupIndex(groups): number | undefined {
```
Comment (Platonn) [inline feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts]: nitpick: missing type of argument

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+        .subscribe((isDesktop) => {
+          if (isDesktop) {
+            selector = 'main' + selector;
+          } else {
+            selector = '.navigation' + selector;
+          }
+        });
+      return document.querySelectorAll(selector);
+    }
+  }
```
Comment (Platonn) [inline feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts]: There are a couple of issues with above implementation:
- our component (service) looks outside the component's DOM by performing lookup on the whole HTML document. Then such a component is harder to reason about. Instead we should lookup only the component's content.
- we're adding conditions for the location of the component in the whole document (in mobile it's different than in desktop)

Instead, we could use [Angular's `@ViewChildren`](https://angular.io/api/core/ViewChildren) which [lookups](https://angular.io/api/core/QueryList) only the certain component's HTML template by the `#templateVariable` marker. For example:


```ts
  // component TS logic (not service provided in root!)
  @ViewChildren('group') groups: QueryList<ElementRef>;
```

```html
<button
    ....
     #group
    ....                  

<button
    ....
     #group
    ....                  
```

File: feature-libs/product-configurator/rulebased/components/form/configurator-form.component.html [hunk 10]
```diff
@@ -1,134 +1,136 @@
...
-      ></cx-configurator-conflict-description>
-    </ng-container>
-    <div
-      class="cx-group-attribute"
-      *ngFor="
-        let attribute of currentGroup.attributes;
-        let indexOfAttribute = index
-      "
-    >
+    <div id="{{ currentGroup.id }}" role="tabpanel">
```
Comment (Platonn) [inline feature-libs/product-configurator/rulebased/components/form/configurator-form.component.html]: Do we _really_ need to introduce global html ID in this place?

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.html [hunk 11]
```diff
@@ -1,66 +1,60 @@
...
+              [cxFocus]="{ autofocus: true }"
+              (click)="navigateUp()"
+              (keydown)="switchGroupOnArrowPress($event, groupIndex, group)"
+            >
+              <cx-icon [type]="iconTypes.CARET_LEFT"></cx-icon>
+              {{ 'configurator.button.back' | cxTranslate }}
+            </button>
+          </ng-container>
+          <button
+            id="{{ group.id }}"
```
Comment (Platonn) [inline feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.html:23]: Do we _really_ need to introduce global html ID here?
(same for               id="back-button")

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts (hunk 2 above)
Comment (Platonn) [reply feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts]: Update: I hope we can get rid of this service at all, see comment: https://github.com/SAP/spartacus/pull/12348#discussion_r638070838

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.html (hunk 4 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.html]: Unfortunately, here it is not possible to get rid of the group ID. For instance, by using Arrow Up or Arrow Down keys another group element is focused in the group menu. There is a difference between an active group and a focused one. The active one is a group that is currently selected. The focused one is currently focused in the group menu, namely gets a blue focus frame. Without knowing ID of the currently focused group element, it will not be possible to find out the right next or previous group to focus.

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.html (hunk 11 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.html:23]: Here it is possible! Good point!

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts (hunk 5 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts]: As far as I understand `isPlatformBrowser(this.platformId)` is needed in case of the server-side rendering. By navigating into the subgroups, the group menu will be rerendered. The same happens when one navigates from the subgroup to the parent group.

Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts]: I am not sure that I understand what you mean here:

> I hope we're safe in SSR, when you use this.document (Angular Injectable DOCUMENT wrapper) instead of the global document object.

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts (hunk 2 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts]: Good hint!  I removed `?` after breakpointService

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts (hunk 8 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts]: Done!

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts (hunk 9 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts]: We had the same dilemma as we thought how to implement our another helper service: `configurator-storefront-utils.service.ts` that takes care of scrolling or focusing the first attribute within the group.
As far as I know we decided us against `@ViewChildren` beacuse there is no option to get an active element. Please, correct me if I am wrong.

File: feature-libs/product-configurator/rulebased/components/form/configurator-form.component.html (hunk 10 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/form/configurator-form.component.html]: According to the [w3.org documentation ](https://www.w3.org/TR/2019/NOTE-wai-aria-practices-1.1-20190814/examples/tabs/tabs-2/tabs.html)regarding groups / tabs navigation, the ID is required for the screen reader, namely  for:
1. `aria-controls` requires `ID` to refer to the group panel element associated with the group and 
2. `aria-labelledby` refers to the group element that controls the group and provides an accessible name for the group panel

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.ts (hunk 3 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.ts]: Done! I took your feedback into account regarding LTR / RTL direction

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts (hunk 6 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts]: If the current index ID is not equal focused index ID then the current index ID will be updated / changed to the focused index ID, otherwise the current index ID stays unchanged.

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.ts (hunk 7 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.ts]: Done!

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts (hunk 6 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts]: I added the JSDocs to all methods in the service

File: feature-libs/product-configurator/rulebased/styles/_configurator-group-menu.scss [hunk 12]
```diff
@@ -1,5 +1,5 @@
 %cx-configurator-group-menu {
-  ul {
+  .cx-group-menu {
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/styles/_configurator-group-menu.scss:2]: We need to document the changes in `/docs/migration/css-changes-in-version-4.md `

File: feature-libs/product-configurator/rulebased/components/form/configurator-form.component.spec.ts [hunk 13]
```diff
@@ -397,4 +397,18 @@ describe('ConfigurationFormComponent', () => {
 
     expect(configuratorCommonsService.updateConfiguration).toHaveBeenCalled();
   });
+
+  describe('createGroupId', () => {
+    it('should return empty string because groupID is null', () => {
+      expect(createComponent().createGroupId(null)).toBeUndefined();
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/form/configurator-form.component.spec.ts:406]: Please don't create new strict mode issues

File: feature-libs/product-configurator/rulebased/components/form/configurator-form.component.spec.ts [hunk 14]
```diff
@@ -397,4 +397,18 @@ describe('ConfigurationFormComponent', () => {
...
+    });
+
+    it('should return empty string because groupID is undefined', () => {
+      expect(createComponent().createGroupId(undefined)).toBeUndefined();
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/form/configurator-form.component.spec.ts:410]: Please don't create new strict mode issues

File: feature-libs/product-configurator/rulebased/components/form/configurator-form.component.ts [hunk 15]
```diff
@@ -96,4 +96,10 @@ export class ConfiguratorFormComponent implements OnInit {
   isConflictGroupType(groupType: Configurator.GroupType): boolean {
     return this.configuratorGroupsService.isConflictGroupType(groupType);
   }
+
+  createGroupId(groupId: string): string {
+    if (groupId) {
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/form/configurator-form.component.ts]: If you really want to support undefined here, please have `createGroupId(groupId?: string)`

File: feature-libs/product-configurator/rulebased/components/form/configurator-form.component.ts [hunk 16]
```diff
@@ -96,4 +96,10 @@ export class ConfiguratorFormComponent implements OnInit {
   isConflictGroupType(groupType: Configurator.GroupType): boolean {
     return this.configuratorGroupsService.isConflictGroupType(groupType);
   }
+
+  createGroupId(groupId: string): string {
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/form/configurator-form.component.ts]: Strict mode issue (Method does not always return something)

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.spec.ts [hunk 17]
```diff
@@ -0,0 +1,300 @@
+import { TestBed } from '@angular/core/testing';
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.spec.ts:2]: Please fix all strict mode issues

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts [hunk 18]
```diff
@@ -0,0 +1,160 @@
+import { Inject, Injectable, PLATFORM_ID } from '@angular/core';
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts]: Please fix all strict mode issues

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.ts [hunk 19]
```diff
@@ -81,19 +87,14 @@ export class ConfiguratorGroupMenuComponent {
...
     protected configRouterExtractorService: ConfiguratorRouterExtractorService,
-    protected configUtils: ConfiguratorStorefrontUtilsService
+    protected configUtils: ConfiguratorStorefrontUtilsService,
+    protected configGroupMenuService: ConfiguratorGroupMenuService,
+    protected directionService: DirectionService
   ) {}
+
   /**
-   * Prevents page down behaviour when users press space key to select buttons
+   * @deprecated since 4.0
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.ts]: We need to introduce schematics for that and delete obsolete methods

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.ts [hunk 20]
```diff
@@ -127,12 +128,14 @@ export class ConfiguratorGroupMenuComponent {
   }
 
   /**
+   * @deprecated since 4.0
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.ts]: Introduce schematics

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.ts [hunk 21]
```diff
@@ -290,4 +291,74 @@ export class ConfiguratorGroupMenuComponent {
       })
     );
   }
+
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.ts:275]: Please fix all strict mode issues in new coding!

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts (hunk 1 above)
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts]: If changes needed: We need to have schematics for that

**Review by ChristophHi — CHANGES_REQUESTED** [review]

File: feature-libs/product-configurator/rulebased/styles/_configurator-group-menu.scss (hunk 12 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/styles/_configurator-group-menu.scss:2]: Done!

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.ts (hunk 19 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.ts]: Done!

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.ts (hunk 20 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.ts]: Done!

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts (hunk 1 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts]: Done!

File: feature-libs/product-configurator/rulebased/components/form/configurator-form.component.html (hunk 10 above)
Comment (Platonn) [reply feature-libs/product-configurator/rulebased/components/form/configurator-form.component.html]: Are we already referring any of those tab panels IDs with `aria-controls`or `aria-labelledby`? If not, do we plan to?

What are possible values of the `currentGroup.id` ? I'm asking just to make sure it's unlikely we'll have value clash with other ID of the same value in the page.

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.html (hunk 4 above)
Comment (Platonn) [reply feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.html]: Thanks, I get it.

I'm just hinting on potential name clashes of thee `id` html attribute in other parts of page. That's why I'd advise to avoid using native `id` attribute or make it enough unique to avoid name clashes.

Do we _really_ need to use the the `id` html attribute for it? What about using a custom `data-*` attribute for it? For example `data-group-id`:
```html
<button [attr.data-group-id]="'back-button'">
...
<button [attr.data-group-id]="group.id">
```

File: feature-libs/product-configurator/rulebased/components/form/configurator-form.component.html (hunk 10 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/form/configurator-form.component.html]: ` <div id="{{ createGroupId(currentGroup.id) }}" role="tabpanel">` was updated ;)
Possible **group ID** looks like: `3-CONF_HT_AUDIO_SYSTEM@Front_Speakers-group`

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts (hunk 5 above)
Comment (Platonn) [reply feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts]: SSR breaks only when you're using globals `document` or `window` - only then you need a wrapper check `isPlatformBrowser`. 

If you're using Angular DOCUMENT object (`this.document` instead of global `document`), SSR won't crash.

If you need to use global window, I recommend WindowRef.nativeWindow - wrapper done by Spartacus. It doesn't crash in SSR, but just return undefined.

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts (hunk 9 above)
Comment (Platonn) [reply feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts]: I'm not sure I understand the problem of using `@ViewChildren` in `ConfiguratorGroupMenuComponent`? When it returns the list of the group elements, you should be able to filter them by any criteria (i.e. filter only if it's native element is focused or active). 

Can you please try using `@ViewChildren` in the component (not the service - it won't be available there)?

File: feature-libs/product-configurator/rulebased/components/form/configurator-form.component.html (hunk 10 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/form/configurator-form.component.html]: `aria-controls` is already supported (see `configurator-group-menu.component.html` line 29)

Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/form/configurator-form.component.html]: I would suggest to add `aria-labelledby` late on as soon as the screen reader comes in play

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.html (hunk 4 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.html]: For `back-button` I removed ID. For other groups, the group ID looks like: `1-CONF_HOME_THEATER_ML@_GEN`

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts [hunk 22]
```diff
@@ -62,22 +61,22 @@ export class ConfiguratorGroupMenuService {
...
-   * @protected
    */
   protected updateCurrentGroupIndex(
     currentGroupIndex: number,
-    focusedGroupIndex: number
+    focusedGroupIndex?: number
   ): number {
-    return focusedGroupIndex !== currentGroupIndex
-      ? focusedGroupIndex
+    return focusedGroupIndex
```
Comment (Platonn) [inline feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts]: Just to doouble check: Are you aware that this statement is falsy also when the index is 0? Taking it into account, is this logic correct?

Btw. this code is not too much readable. Maybe we can improve by adding some `if` and then ternary expression

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts [hunk 23]
```diff
@@ -12,11 +12,11 @@ import { Configurator } from '../../core/model/configurator.model';
   providedIn: 'root',
 })
 export class ConfiguratorStorefrontUtilsService {
-  // TODO(#11681): make keyboardFocusService a required dependency and remove deprecated constructor
   constructor(
     protected configuratorGroupsService: ConfiguratorGroupsService,
     @Inject(PLATFORM_ID) protected platformId: any,
-    protected keyboardFocusService?: KeyboardFocusService
+    @Inject(DOCUMENT) protected document: any,
```
Comment (Platonn) [inline feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts]: Why do we add dependency on DOCUMENT here? I cannot see it's used anywhere in this class

Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts:18]: I could repace both dependencies: 
```
@Inject(PLATFORM_ID) protected platformId: any,
@Inject(DOCUMENT) protected document: any
```
through `windowRef: WindowRef` as we discussed yesterday.
Is it fine for you?

File: feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts (hunk 22 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.service.ts]: Took into account your feedback: 
- add `if` condition for readability,
- add some tests.

The **current group index** will be updated if is is not equal to the **focused group index**.
If the **focused group index** is `null`, `undefine` or `equal` to the **current group index**, nothing happens

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts (hunk 23 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts]: As discussed, I replaced both dependencies:
```
@Inject(PLATFORM_ID) protected platformId: any,
@Inject(DOCUMENT) protected document: any
```
through `WindowRef` wrapper. 
Furthermore, the constructor deprecation for `ConfiguratorStorefrontUtilsService` was updated accordingly.

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts [hunk 24]
```diff
@@ -92,7 +93,7 @@ export class ConfiguratorStorefrontUtilsService {
     if (element instanceof HTMLElement) {
       topOffset = element.offsetTop;
     }
-    window.scroll(0, topOffset);
+    this.windowRef.nativeWindow.scroll(0, topOffset);
```
Comment (Platonn) [inline feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts]:
```suggestion
    this.windowRef.nativeWindow?.scroll(0, topOffset);
```

native window might be undefined (mostly in SSR) + this likely fixes strict typing issue

**Review by Platonn — DISMISSED** [review]

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts (hunk 24 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts]: Done!

**Review by ChristophHi — DISMISSED** [review]

File: projects/schematics/src/migrations/4_0/constructor-deprecations/data/configurator-form.component.migration.ts [hunk 25]
```diff
@@ -0,0 +1,19 @@
...
+  CONFIGURATOR_STOREFRONT_UTILS_SERVICE,
+  SPARTACUS_PRODUCT_CONFIGURATOR_RULEBASED,
+} from '../../../../shared/constants';
+import { ConstructorDeprecation } from '../../../../shared/utils/file-utils';
+
+export const CONFIGURATOR_FORM_COMPONENT_MIGRATION: ConstructorDeprecation = {
+  //feature-libs/product-configurator/rulebased/components/form/configurator-form.component.ts
+  class: CONFIGURATOR_FORM_COMPONENT,
+  importPath: SPARTACUS_PRODUCT_CONFIGURATOR_RULEBASED,
+  deprecatedParams: [],
```
Comment (ChristophHi) [inline projects/schematics/src/migrations/4_0/constructor-deprecations/data/configurator-form.component.migration.ts]: This has to be specified, otherwise the update mechanism won't pick the class. Did you unit-test this migration?

File: projects/schematics/src/migrations/4_0/constructor-deprecations/data/configurator-storefront-utils.service.migration.ts [hunk 26]
```diff
@@ -23,7 +25,17 @@ export const CONFIGURATOR_STOREFRONT_UTILS_SERVICE_MIGRATION: ConstructorDepreca
       importPath: ANGULAR_CORE,
     },
   ],
+  removeParams: [
```
Comment (ChristophHi) [inline projects/schematics/src/migrations/4_0/constructor-deprecations/data/configurator-storefront-utils.service.migration.ts:28]: Did you unit test this migration?

File: projects/schematics/src/migrations/4_0/methods-and-properties-deprecations/data/configurator-group-menu.component.migration.ts [hunk 27]
```diff
@@ -0,0 +1,25 @@
...
+  TODO_SPARTACUS,
+} from '../../../../shared/constants';
+
+import { MethodPropertyDeprecation } from '../../../../shared/utils/file-utils';
+
+//feature-libs/product-configurator/rulebased/components/group-menu/configurator-group-menu.component.ts
+export const CONFIGURATOR_GROUP_MENU_COMPONENT_MIGRATION: MethodPropertyDeprecation[] = [
+  {
+    class: CONFIGURATOR_GROUP_MENU_COMPONENT,
+    importPath: SPARTACUS_PRODUCT_CONFIGURATOR,
```
Comment (ChristophHi) [inline projects/schematics/src/migrations/4_0/methods-and-properties-deprecations/data/configurator-group-menu.component.migration.ts]: No, this needs to be imported from /rulebased endpoint

File: projects/schematics/src/migrations/4_0/constructor-deprecations/constructor-deprecations.ts [hunk 28]
```diff
@@ -130,11 +131,11 @@ export const CONSTRUCTOR_DEPRECATION_DATA: ConstructorDeprecation[] = [
   SAVED_CART_DETAILS_OVERVIEW_COMPONENT_MIGRATION_V2,
   DYNAMIC_ATTRIBUTE_SERVICE_MIGRATION,
   WINDOW_REF_MIGRATION,
-  CONFIGURATOR_CART_ENTRY_INFO_COMPONENT_MIGRATION,
```
Comment (ChristophHi) [inline projects/schematics/src/migrations/4_0/constructor-deprecations/constructor-deprecations.ts:135]: Why did you remove that? Please re-insert

**Review by ChristophHi — CHANGES_REQUESTED** [review]

File: projects/schematics/src/migrations/4_0/constructor-deprecations/constructor-deprecations.ts (hunk 28 above)
Comment (Larisa-Staroverova) [reply projects/schematics/src/migrations/4_0/constructor-deprecations/constructor-deprecations.ts:135]: **CONFIGURATOR_CART_ENTRY_INFO_COMPONENT_MIGRATION** is there at another line could be found (see line 153)

File: projects/schematics/src/migrations/4_0/constructor-deprecations/data/configurator-form.component.migration.ts (hunk 25 above)
Comment (Larisa-Staroverova) [reply projects/schematics/src/migrations/4_0/constructor-deprecations/data/configurator-form.component.migration.ts]: Here nothing was deprecated, new dependency was added

File: projects/schematics/src/migrations/4_0/methods-and-properties-deprecations/data/configurator-group-menu.component.migration.ts (hunk 27 above)
Comment (Larisa-Staroverova) [reply projects/schematics/src/migrations/4_0/methods-and-properties-deprecations/data/configurator-group-menu.component.migration.ts]: Done!

File: projects/schematics/src/migrations/4_0/constructor-deprecations/data/configurator-form.component.migration.ts (hunk 25 above)
Comment (ChristophHi) [reply projects/schematics/src/migrations/4_0/constructor-deprecations/data/configurator-form.component.migration.ts]: still the existing constructor attributes need to be stated, otherwise the class is not picked. Please check with a unit test (compare with my examples from the FB...)

Comment (Larisa-Staroverova) [reply projects/schematics/src/migrations/4_0/constructor-deprecations/data/configurator-form.component.migration.ts]: Done!

File: projects/schematics/src/migrations/4_0/constructor-deprecations/data/configurator-storefront-utils.service.migration.ts (hunk 26 above)
Comment (Larisa-Staroverova) [reply projects/schematics/src/migrations/4_0/constructor-deprecations/data/configurator-storefront-utils.service.migration.ts:28]: Done!

**Review by ChristophHi — APPROVED** [review]

**Review by marlass — APPROVED** [review]
